### PR TITLE
[TT-5375] Add is_oas flag to APIDefinition

### PR DESCRIPTION
**Pull request by jeffy-mathew**
## Description
Add is_oas flag in APIDefinition to distinguish whether API is created/migrated to OAS spec

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [ ] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [x] `go fmt -s`
  - [x] `go vet`

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit:  
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2375795240)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: 76ea061ddd668bc006048dcf14bb28ce71ca49ff 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2375830809)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: 32301bd00faaf15607917e3d97a485eabcd6af80 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2375849300)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: 8d98c786c9f26f53123a474a356612a97dc7526d 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2375922231)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: bf4f97d7d05de69a588c2ce489c838a444177a9c 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2375994488)

File: gateway/api.go [hunk 1]
```diff
@@ -965,34 +966,44 @@ func (gw *Gateway) handleGetAPIOAS(apiID string, modePublic bool) (interface{},
 
 }
 
-func (gw *Gateway) handleAddOrUpdateApi(apiID string, r *http.Request, fs afero.Fs, oasTyped bool) (interface{}, int) {
+func (gw *Gateway) handleAddOrUpdateApi(apiID string, r *http.Request, fs afero.Fs, oasAPI bool) (interface{}, int) {
```
Comment (furkansenharputlu) [inline gateway/api.go:969]: oasEndpoint

File: gateway/api.go [hunk 2]
```diff
@@ -965,34 +966,44 @@ func (gw *Gateway) handleGetAPIOAS(apiID string, modePublic bool) (interface{},
...
+
+		if err := json.NewDecoder(r.Body).Decode(&oasObj); err != nil {
 			log.Error("Couldn't decode new OAS object: ", err)
 			return apiError("Request malformed"), http.StatusBadRequest
 		}
 
-		oas.ExtractTo(&newDef)
+		oasObj.ExtractTo(&newDef)
 	} else {
 		if err := json.NewDecoder(r.Body).Decode(&newDef); err != nil {
```
Comment (furkansenharputlu) [inline gateway/api.go:998]:
```
if !isCreate && spec.IsOAS && !oasEndpoint{
```

File: gateway/api.go [hunk 3]
```diff
@@ -965,34 +966,44 @@ func (gw *Gateway) handleGetAPIOAS(apiID string, modePublic bool) (interface{},
...
 	if gw.GetConfig().UseDBAppConfigs {
 		log.Error("Rejected new API Definition due to UseDBAppConfigs = true")
 		return apiError("Due to enabled use_db_app_configs, please use the Dashboard API"), http.StatusInternalServerError
 	}
 
 	var newDef apidef.APIDefinition
-	var oas oas.OAS
+	var oasObj oas.OAS
+
+	isCreate := r.Method == http.MethodPost
```
Comment (furkansenharputlu) [inline gateway/api.go:978]: instead of checking every time, we can split handlers like `handleAddAPI` and `handleUpdateAPI`

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: 7c56bad55d1dcde3c3866706ec3ee345f0915685 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2377526981)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: 00f30e80c39002f5c5439fe3beb581e5bbae2c96 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2377546942)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: e47ccac4c5d0e62b0e24837fee1487b4382227e4 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2377769638)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: 70c883b1bb8502b4aa53471c5e86a92c16257631 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2377676255)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: e4cd486ecb9d1228a8bd0e44049e5048ef752f84 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2378202619)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: 41f11d7ed5f602bd4357c1d9ae7abf15b1d6e1f7 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2378257394)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: 083d91c95ac831a2912f903a5bc86e699d0de214 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2379131744)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: 695a689d779b71837981841c330b331c391a2688 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2379358371)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: bb817f6ea1c6b08f16be6817711cee16a81ff5fc 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2379563536)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: 7a700bfa0a7f8d9f62397675a1189871f17c9d69 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2381952656)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: 54d83a5d401f9edce11a1ce5061a00784b3a19bf 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2382063664)

**Review by urbanishimwe — APPROVED** [review]

**Review by furkansenharputlu — APPROVED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4067/merge
Commit: 661600a0620c3bd1db74d806264ead27fa6f757e 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2382681253)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/heads/master
Commit: 1d64e7b45a74e4de068c1700da058003ed005438 [TT-5375] Add is_oas flag to APIDefinition (#4067)

* add is_oas flag to APIDefinition

[changelog]
added: is_oas flag in APIDefinition.
Triggered by: push (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2382777309)

**Pull request merged**